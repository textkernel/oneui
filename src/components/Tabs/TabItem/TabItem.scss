.TabItem {
    color: var(--color-neutral-80);
    cursor: pointer;
    outline: none;
    padding: var(--space-75) var(--space-150) var(--space-100) var(--space-150);
    text-decoration: none;
    transition: all var(--transition-duration);
    text-align: center;
    border-bottom: 4px solid transparent;
    font-weight: var(--font-weight-bold);


    &:hover,
    &:focus {
        border-bottom: 4px solid var(--color-background);
        background-color: var(--color-neutral-10);
        border-radius: var(--border-radius);
    }

    &--isActive,
    &:active {
        color: var(--color-brand-50);
        border-bottom: 4px solid var(--color-brand-50);
        background-color: var(--color-background);
        border-radius: 0;
    }

    &--isActive {
        &:hover,
        &:focus {
            color: var(--color-brand-50);
            border-bottom: 4px solid var(--color-brand-50);
            background-color: var(--color-highlight-background);
            border-radius: 0;
        }
    }

    &--isBlock {
        flex-grow: 1;
    }

    &--disabled {
        cursor: not-allowed;
        color: var(--color-neutral-30);

        &:hover,
        &:focus {
            border-bottom: 4px solid transparent;
            color: var(--color-neutral-30);
        }
    }
}
