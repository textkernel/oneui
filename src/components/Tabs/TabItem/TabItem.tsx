import * as React from 'react';
import { bem } from '../../../utils';
import { ENTER_KEY } from '../../../constants';
import styles from './TabItem.scss';
import { ReactNode } from '../../../customTypes/types';

export interface Props<T> extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onSelect'> {
    /** Id of this tab */
    tabId: T;
    /** Renders an active tab. This prop will be set by TabsBar if activeTabId is defined there */
    isActive?: boolean;
    /** Disabled tab item */
    disabled?: boolean;
    /** A callback  when the tab is clicked. It will not be called for active or disabled tabs. This prop will be set by TabsBar if onSelect is defined there */
    onSelect?: (tabId: T) => void;
    /** Label of the tab. Expected to be a string like node. E.g. `label` or `label <span>(3)</span>`  */
    children: ReactNode;
    /** used for styling when TabsBar is full width. Will be set by TabsBar, no need to set manually */
    isBlock?: boolean;
}

const { block } = bem('TabItem', styles);

export function TabItem<T = string | number>({
    tabId,
    isActive = false,
    onSelect,
    disabled = false,
    isBlock = false,
    children,
    ...rest
}: Props<T>) {
    const handleClick = () => {
        if (!isActive && !disabled && onSelect) {
            onSelect(tabId);
        }
    };

    const handleKeyDown = (e: KeyboardEvent) => {
        if (e.key === ENTER_KEY) {
            handleClick();
        }
    };

    return (
        <div
            tabIndex={0}
            {...rest}
            {...block({ isActive, isBlock, ...rest })}
            role="tab"
            aria-selected={isActive}
            onClick={handleClick}
            onKeyDown={handleKeyDown}
        >
            {children}
        </div>
    );
}

TabItem.displayName = 'TabItem';
