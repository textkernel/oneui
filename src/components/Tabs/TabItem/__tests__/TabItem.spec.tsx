import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { TabItem } from '../TabItem';

describe('<TabItem>', () => {
    const onSelectMock = jest.fn();

    let view: RenderResult;

    it('should render a tab item correctly with minimum props', () => {
        view = render(<TabItem tabId="1">Tab label</TabItem>);

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('tab', { name: 'Tab label' })).toBeInTheDocument();
    });

    it('should render correctly with all props', () => {
        view = render(
            <TabItem tabId="1" isBlock isActive onSelect={onSelectMock}>
                Tab label <span style={{ color: 'grey' }}>(3)</span>
            </TabItem>
        );

        const tab = screen.getByRole('tab', { name: 'Tab label (3)' });

        expect(view.container).toMatchSnapshot();
        expect(tab).toBeInTheDocument();
        expect(tab).toHaveAttribute('aria-selected', 'true');
    });

    describe('callbacks', () => {
        it('should call onSelect on simple tab', async () => {
            const user = userEvent.setup();
            view = render(
                <TabItem tabId="1" onSelect={onSelectMock}>
                    Tab label
                </TabItem>
            );

            await user.click(screen.getByRole('tab'));

            expect(onSelectMock).toHaveBeenCalledTimes(1);
        });

        it('should not call onSelect on active tab', async () => {
            const user = userEvent.setup();
            view = render(
                <TabItem tabId="1" isActive onSelect={onSelectMock}>
                    Tab label
                </TabItem>
            );

            await user.click(screen.getByRole('tab'));

            expect(onSelectMock).toHaveBeenCalledTimes(0);
        });

        it('should not call onSelect on disabled tab', async () => {
            const user = userEvent.setup();
            view = render(
                <TabItem tabId="1" disabled onSelect={onSelectMock}>
                    Tab label
                </TabItem>
            );

            await user.click(screen.getByRole('tab'));

            expect(onSelectMock).toHaveBeenCalledTimes(0);
        });
    });
});
