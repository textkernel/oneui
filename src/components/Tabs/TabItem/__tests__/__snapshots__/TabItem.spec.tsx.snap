// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabItem> should render a tab item correctly with minimum props 1`] = `
<div>
  <div
    aria-selected="false"
    class="TabItem"
    role="tab"
    tabindex="0"
  >
    Tab label
  </div>
</div>
`;

exports[`<TabItem> should render correctly with all props 1`] = `
<div>
  <div
    aria-selected="true"
    class="TabItem TabItem--isActive TabItem--isBlock"
    role="tab"
    tabindex="0"
  >
    Tab label 
    <span
      style="color: grey;"
    >
      (3)
    </span>
  </div>
</div>
`;
